memgpt/agent.py

```diff
@@ -104,22 +109,34 @@ def initialize_memory(ai_notes, human_notes):
 
 
 def construct_system_with_memory(system, memory, memory_edit_timestamp, archival_memory=None, recall_memory=None, include_char_count=True):
-    full_system_message = "\n".join(
-        [
-            system,
-            "\n",
-            f"### Memory [last modified: {memory_edit_timestamp.strip()}]",
-            f"{len(recall_memory) if recall_memory else 0} previous messages between you and the user are stored in recall memory (use functions to access them)",
-            f"{len(archival_memory) if archival_memory else 0} total memories you created are stored in archival memory (use functions to access them)",
-            "\nCore memory shown below (limited in size, additional information stored in archival / recall memory):",
-            f'<persona characters="{len(memory.persona)}/{memory.persona_char_limit}">' if include_char_count else "<persona>",
-            memory.persona,
-            "</persona>",
-            f'<human characters="{len(memory.human)}/{memory.human_char_limit}">' if include_char_count else "<human>",
-            memory.human,
-            "</human>",
-        ]
-    )
+    if isinstance(memory, InContextMemory):
+        full_system_message = "\n".join(
+            [
+                system,
+                "\n",
+                f"### Memory [last modified: {memory_edit_timestamp.strip()}]",
+                f"{len(recall_memory) if recall_memory else 0} previous messages between you and the user are stored in recall memory (use functions to access them)",
+                f"{len(archival_memory) if archival_memory else 0} total memories you created are stored in archival memory (use functions to access them)",
+                "\nCore memory shown below (limited in size, additional information stored in archival / recall memory):",
+                f'<persona characters="{len(memory.persona)}/{memory.persona_char_limit}">' if include_char_count else "<persona>",
+                memory.persona,
+                "</persona>",
+                f'<human characters="{len(memory.human)}/{memory.human_char_limit}">' if include_char_count else "<human>",
+                memory.human,
+                "</human>",

```

Comment:
@Maximilian-Winter if we're trending in the direction of making the layout of the entire system block dynamic (which I love btw, this is an awesome PR), do you think it makes sense to also go the extra step and make the "template" of the entire system block including the memory dynamic?

Eg right now you have approximately the following in the first system message:
```
{system message, can be dynamically created}
{memory block}
```

where `{memory block}` is
```
### Memory [last modified X]
{human memory}
{persona memory}
{or custom memory}
```

but the `"### Memory [last modified: {memory_edit_timestamp.strip()}]` part is still hardcoded, whereas this could also be rolled into a template maybe? e.g. if people want to play with using other things than xml tags?

Or maybe this is better off just being added as a modification to your configurable core memory code?

Reply:
This is probably overkill, but something like "memory_edit_timestamp.strip()" could even be pulled out into a templating variable called something like `{CURRENT_TIME}`

It may not work properly given how we pass `memory_edit_timestamp` around (I'd need to take a look), since you'd be relying on the timestamp at time of recreating the system prompt to be correct as a timestamp for editing memory (tricky if you're recompiling the other parts of the system prompt), so this particular example may not make much sense, just something to consider.

Reply:
I actually think these additions all would make sense. And implemented them.